[feat, ui] 메인 화면 뷰 그리기

kelly-chui opened the pull request:
### 🔖  Issue Number

close #2 

### 📙 작업 내역

> 구현 내용 및 작업 했던 내역을 적어주세요.
> 
- [X]  메인 홈 화면 그리기
- [X]  메인 홈 화면 프로필 뷰 그리기

<img src = "https://github.com/user-attachments/assets/25fa6cda-2cdc-4487-9379-715be1811c40" width = 300>

### 📝 PR 특이사항 
![IMG_0379](https://github.com/user-attachments/assets/072f83ff-c64b-4a2c-8776-f323cf00b774)

> PR을 볼 때 팀원에게 알려야 할 특이사항을 알려주세요.
> 
- 프로필이 DB 혹은 네트워크에서 로드되기전에 뷰를 어떻게 그릴것인지
    1. 일단 로드 안된 상태로 깡통이라도 보여주기
    2. `lazy` 프로퍼티로 만들어서 다 로드 되면 그리고 `addSubView()` 하기
    [HIG보면 1번이 국룰인거 같아서 1번으로](https://developer.apple.com/design/human-interface-guidelines/loading)

- 처음부터 모든 프로필 정보를 다 받아오느냐? 혹은 프로필 카드를 뒤집었을 때 받아오느냐?
    - 처음부터 다 받아오면 DB에 요청을 2번 할 필요가 없어짐
    - 뒤집었을 때 받아오면 요청을 2번 해야함, 대신 안뒤집는 행동을 했을 때 쓸모없는 데이터를 받아오지 않을 수 있음
    - 중요한건 프로필 정보는 네트워크보단 DB에서 받아온다는 것 첫 번째가 더 나을거 같음

### 👻 레퍼런스
- [내가 공부한 VIPER 패턴](https://prickly-theater-8f3.notion.site/VIPER-139ade8f37658032880ae515827a8606?pvs=4)

Commit:
[feat] #2 MainTabBarFactory 추가, 애플리케이션의 탭 바를 생성합니다.

Commit:
[feat] #2 MainViewController.swift를 추가했습니다. 유저는 이곳에서 프로필을 확인할 수 있습니다.

Commit:
[feat] #2 ProfileCardView 생성, 이 뷰는 사용자의 프로필을 직접 표시합니다.

Commit:
[chore] 프로젝트 파일 업데이트

Commit:
[chore] #2 탭바 관련 코드를 제거했습니다. 이슈를 따로 분리할 계획입니다.

Commit:
[ui] #2 프로필 카드 뷰의 editButton의 크기와 색상을 조절하고, 기본 설정된 프로필 이름과 키워드를 제거했습니다

Commit:
[fix] #2 ViewController 네이밍 수정 MainViewController -> HomeView

Commit:
[ui] #2 ProfilCardView에 그림자를 적용했습니다

green-yoon87 commented on SniffMeet/SniffMeet/Source/Home/Main/View/ProfileCardView.swift:
```diff
@@ -22,9 +22,15 @@ final class ProfileCardView: UIView {
         return label
     }()
     let editButton: UIButton = {
```
P3. 뷰 속성들을 private 접근제어자를 이용해서 캡슐화하는게 좋아보여요!

kelly-chui replied:
👍

green-yoon87 commented on SniffMeet/SniffMeet/Source/Home/Main/View/ProfileCardView.swift:
```diff
@@ -108,4 +109,12 @@ final class ProfileCardView: UIView {
     func setProfileImage(profileImage: UIImage) {
         profileImageView.image = profileImage
     }
+    func setShadow() {
```
P3. shadow를 설정하셨군요! UIView extension에 해당 함수를 작성하면 더 범용성 있게 사용할 수도 있겠네요

kelly-chui replied:
이미 뷰에 내장된 레이어 단위로 접근할 수 있는 거라서 익스텐션까지는 필요하지 않다고 생각해요!

green-yoon87 commented on SniffMeet/SniffMeet/Source/Home/Main/View/ProfileCardView.swift:
```diff
@@ -0,0 +1,103 @@
+        button.layer.cornerRadius = 22
+        return button
+    }()
+    lazy var keywordStack: UIStackView = {
```
P3. imageView는 ~ImageView라고 명명하셨는데 StackView는 ~Stack이라고 명명하셨네요. 
일관성 있게 네이밍하시면 더 좋을 거 같습니당 

P2. Stackview에 lazy 키워드를 작성하신 이유가 뭘까요? lazy 프로퍼티는 thread-safe하지 않습니다.

kelly-chui replied:
1. View를 붙이겠습니다.

2. 스택 뷰 프로퍼티에 작성된 클로저가 평가되는 시점에 해당 뷰 내부의 다른 뷰 프로퍼티를 이용하여 스택을 생성하려면 `lazy` 키워드가 필요합니다. 클로저 평가와 동시에 초기값으로 뷰를 초기화하는 것이 아닌 super.init() 이후에 추가하는 형식으로 옮겨보겠습니다.

green-yoon87 commented on SniffMeet/SniffMeet/Source/Home/Main/View/ProfileCardView.swift:
```diff
@@ -0,0 +1,103 @@
+        nameLabel.layer.shadowRadius = 3.0
+        nameLabel.layer.masksToBounds = false
+    }
+    func setKeywords(keywords: [String]) {
```
P1. keyword entity를 어떻게 구성해야할지 논의가 필요하겠네요!  
저는 개인적으로 enum이 적합하다고 생각했습니당

kelly-chui replied:
데이터를 뷰에서 사용하는 밸류로 변경시키는건 presenter의 역할이라고 생각해서 뷰에선 string을 사용했습니다.

green-yoon87 commented on SniffMeet/SniffMeet/Source/Home/Main/View/ProfileCardView.swift:
```diff
@@ -0,0 +1,103 @@
+            profileInfoStack.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 20)
+        ])
+    }
+    func setName(name: String) {
```
P3. 함수 이름 의도와 다른 역할도 많이 수행을 하는 거 같아요.

kelly-chui replied:
레이블 텍스트가 업데이트 되면 그림자 레이어도 업데이트 되어야 합니다.
1. setNameLabelShadow() 메소드를 만들어서 옮겨보겠습니다.

green-yoon87 reviewed (changes requested):
수고하셨습니당!! P2가 한개 있어서 request change를 선택했습니당

Commit:
[fix] #2 픽스

1. 네이밍 컨벤션에 맞춰서 네이밍을 변경했습니다. HomeViewController → HomeView
2. 뷰 프로퍼티의 액세스 컨트롤을 지정했습니다.
3. setName()함수의 내부 동작을 분리했습니다.
4. StackView 프로퍼티가 생성되는 시점이 아닌 super.init() 이후에 서브 뷰를 추가하도록 수정했습니다.

Commit:
[fix] #2 내부 메소드 정리

1. setKeywordViews와 setKeywords가 서로 다른 점이 없는 메소드라서 합쳤습니다. 기존 setKeywords는 setKeywordViews를 호출만 하는 동작을 했습니다.
2. Configure Atrribute, constraint, Hierarchy 메소드를 만들어서 각각의 역할에 맞는 메소드를 호출하도록 했습니다.

Commit:
[chore] #2 프로젝트 파일 업데이트

kelly-chui committed:
Merge branch 'dev' into feature/#2/add-main-view
